src/components/menu/menu-popover.tsx

```diff
@@ -1,29 +1,70 @@
-import { type FC, type MutableRefObject, useRef } from 'react'
+import { type FC, type MutableRefObject, useLayoutEffect, useRef, useState } from 'react'
 import { useClickOutside } from '../../hooks/use-click-outside'
 import { useMenuContext } from './menu-context'
 import { ElMenuPopover } from './styles'
 
-export const MenuPopover: FC = ({ children }) => {
+export const calculatePopoverPosition = (
+  container: HTMLElement,
+  popover: HTMLDivElement,
+  setPopoverStyle: React.Dispatch<React.SetStateAction<React.CSSProperties>>,
+  additionalGap = 0,
+) => {
+  const triggerBtn = container.querySelector('[role="button"]')
+  if (triggerBtn) {
+    const viewportHeight = window.innerHeight
+    const popoverHeight = popover.getBoundingClientRect().height
+    const { bottom: triggerBottomPos, height: triggerheight } = triggerBtn.getBoundingClientRect()
+    const gap = 3 // the shadow width or .etc
+    const spaceBelowButton = viewportHeight - triggerBottomPos
+
+    const top = popoverHeight > spaceBelowButton ? 0 - popoverHeight - gap : triggerheight + gap
+
+    setPopoverStyle({ top: top + additionalGap })
+  }
+}
+
+export const MenuPopover: FC<{
+  /**
+   * Optional parameter to adjust the vertical position of the popover
+   *
+   * @default 0
+   */
+  additionalGap?: number
```

Comment:
**suggestion:** I think it's nicer to accept a `string` as that allows consumers to specify any valid CSS length value.

Also, I'd suggest a prop name like `yOffset` where the default is the 3px of whatever is specified by the Design System. This way, engineers only need to match the total offset shown in a design instead of needing to manually subtract the default offset.

Reply:
> suggestion: I think it's nicer to accept a string as that allows consumers to specify any valid CSS length value.

I'm not certain about this, as I calculate the available space using DOM methods like getBoundingClientRect(), which returns values in pixels. I expect this prop to be used rarely, as the default y-offset should generally be correct. However, if the consumer requires additional space, this prop can be utilized

and by the way, since the app-bar and bottom-bar has `position: fixed` container, I think I don't have to support this since it(auto anchor) can be overriden by e.g
```css
 .el-menu-popover {
    top: var(--nav-height) !important;
  }
```

> Also, I'd suggest a prop name like yOffset where the default is the 3px of whatever is specified by the Design System. This way, engineers only need to match the total offset shown in a design instead of needing to manually subtract the default offset.

yeah, I think its more make sense, I will note this 👍